effect/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -0,0 +1,547 @@
+// Copyright (C) 2017 John A. De Goes. All rights reserved.
+package scalaz.effect
+
+import scala.annotation.switch
+import scala.concurrent.duration.Duration
+
+import scalaz.data.Disjunction._
+import scalaz.data.Maybe
+
+/**
+ * An `IO[A]` ("Eye-Oh of A") is an immutable data structure that describes an
+ * effectful action that may throw an exception (`Throwable`), run forever, or
+ * produce a single `A` at some point in the future.
+ *
+ * `IO` values are ordinary immutable values, and may be used like any other
+ * values in purely functional code. Because `IO` values just *describe*
+ * effects, which must be interpreted by a separate runtime system, they are
+ * entirely pure and do not violate referential transparency.
+ *
+ * `IO` values can efficiently describe the following classes of effects:
+ *
+ *  * **Pure Values** &mdash; `IO.apply`
+ *  * **Synchronous Effects** &mdash; `IO.sync`
+ *  * **Asynchronous Effects** &mdash; `IO.async`
+ *  * **Concurrent Effects** &mdash; `io.fork`
+ *  * **Resource Effects** &mdash; `io.bracket`
+ *
+ * The concurrency model is based on *fibers*, a user-land lightweight thread,
+ * which permit cooperative multitasking, fine-grained interruption, and very
+ * high performance with large numbers of concurrently executing fibers.
+ *
+ * `IO` values compose with other `IO` values in a variety of ways to build
+ * complex, rich, interactive applications. See the methods on `IO` for more
+ * details about how to compose `IO` values.
+ *
+ * In order to integrate with Scala, `IO` values must be interpreted into the
+ * Scala runtime. This process of interpretation executes the effects described
+ * by a given immutable `IO` value. For more information on interpreting `IO`
+ * values, see the default interpreter in `RTS` or the safe main function in
+ * `SafeApp`.
+ */
+sealed abstract class IO[A] { self =>
+  /**
+   * Maps an `IO[A]` into an `IO[B]` by applying the specified `A => B` function
+   * to the output of this action.
+   */
+  final def map[B](f: A => B): IO[B] = (self.tag : @switch) match {
```

Comment:
I think a note to the effect of "`map` is not always stack-safe" may be helpful.

Reply:
Added note. It's stack safe to > 10k repetitions on the JVM (probably much more).